jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4301,31 +4306,92 @@ ConstantPool getConstantPool() {
+		if (nestHost != this) {
+			SecurityManager securityManager = System.getSecurityManager();
+			if (nestHost == null) {
+				nestHost = this;
+			} else if (securityManager != null) {
+				ClassLoader callerClassLoader = ClassLoader.getCallerClassLoader();
+				ClassLoader nestHostClassLoader = nestHost.getClassLoader();			
+				if (!com.ibm.oti.util.Util.doesClassLoaderDescendFrom(nestHostClassLoader, callerClassLoader)) {
+					String nestHostPackageName = nestHost.getPackageName();
+					if (nestHostPackageName != null) {
+						securityManager.checkPackageAccess(nestHostPackageName);
+					}
+				}
+			}
+		}
+	}
+	return nestHost;
 }
 
 /**
  * Returns true if the class passed has the same nest top as this class.
- * The list of nest members in the classfile is permitted to contain
- * duplicates, or to explicitly include the nest host. It is not required
- * that an implementation of this method removes these duplicates.
  * 
  * @param		that		The class to compare
  * @return		true if class is a nestmate of this class; false otherwise.
  *
  */
 public boolean isNestmateOf(Class<?> that) {
-	return this.getNestHost() == that.getNestHost();
+	Class<?> thisNestHost = this.nestHost;
+	if (null == thisNestHost) {
+		thisNestHost = this.getNestHostImpl();
+	}
+	Class<?> thatNestHost = that.nestHost;
+	if (null == thatNestHost) {
+		thatNestHost = that.getNestHostImpl();
+	}
+	return thisNestHost == thatNestHost;
 }
 
 /**
  * Answers the nest member classes of the receiver's nest host.
  *
+ * @throws		SecurityException if a SecurityManager is present & package access is not allowed
+ * @throws		LinkageError if there is any problem loading or validating a nest member or the nest host
  * @return		the host class of the receiver.
  */
-public Class<?>[] getNestMembers() {
-	return getNestMembersImpl();
+@CallerSensitive
+public Class<?>[] getNestMembers() throws LinkageError, SecurityException {
+	if (null == nestMembers) {
+		nestMembers = getNestMembersImpl();
+	}
+	if (null != nestMembers) {
+		/* The specification requires that if:
+		 *    - a returned class is not the current class
+		 *    - a security manager, s, is present
+		 *    - the caller's class loader is not the same or an ancestor of the returned class
+		 *    - s.checkPackageAccess() disallows access to the package of the returned class
+		 * then throw a SecurityException. Since all classes in the same nest
+		 * must be in the same runtime package and thus the same class loader,
+		 * it is sufficient to check classloader ancestry once.
+		 */
+		SecurityManager securityManager = System.getSecurityManager();
+		if (null != securityManager) {
+			/* All classes in a nest must be in the same runtime package & therefore same classloader */
+			ClassLoader nestMemberClassLoader = this.getClassLoader();
+			ClassLoader callerClassLoader = ClassLoader.getCallerClassLoader();
+			if (!com.ibm.oti.util.Util.doesClassLoaderDescendFrom(nestMemberClassLoader, callerClassLoader)) {
```

Comment:
you can - `import static com.ibm.oti.util.Util.doesClassLoaderDescendFrom;`. then, you can directly call `doesClassLoaderDescendFrom`.